Fixed Test-Modulemanifest to normalize paths correctly before validating

SteveL-MSFT opened the pull request:
changed hard coded Windows directory separator and resolved path so the slashes are correct
added test
addresses #2610

SteveL-MSFT committed:
fixing #2610
changed hard coded Windows directory separator and resolved path so the slashes are correct
added test

lzybkr commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,13 @@
+        "" > testdrive:/module/foo/bar.psm1
+        "" > testdrive:/module/bar/foo.psm1
+        New-ModuleManifest -NestedModules foo\bar.psm1,bar/foo.psm1 -RootModule foo\bar.psm1 -RequiredAssemblies foo/bar.psm1,bar\foo.psm1 -Path testdrive:/module/test.psd1
+        Test-ModuleManifest -Path testdrive:/module/test.psd1 -ErrorAction Stop | Should BeOfType System.Management.Automation.PSModuleInfo
```
Do you want some additional validation on the paths, maybe a `Test-Path` on the `NestedModules`, `RootModule`, and `RequiredAssemblies`.

There are other module properties that specify files, maybe you should be testing those too - `TypesToProcess`, `FormatsToProcess`, `ScriptsToProcess`, maybe `FileList` and `ModuleList`?

I wonder if it's a bad idea to specify `RequiredAssemblies` without a valid assembly - if that's not already causing problems, I can see it causing problems in the future.

SteveL-MSFT replied:
The semantics of test-modulemanifest seems to only verify existence of files and not if they are valid.  Rather than verify the files exist and being validated, I'm adding a negative test where test-modulemanifest should fail if any paths don't exist

lzybkr commented on src/System.Management.Automation/engine/Modules/TestModuleManifestCommand.cs:
```diff
@@ -288,10 +288,19 @@ private bool IsValidFilePath(string path, PSModuleInfo module, bool verifyPathSc
+                CmdletProviderContext cmdContext = new CmdletProviderContext(this);
+                Collection<PathInfo> pathInfos = SessionState.Path.GetResolvedPSPathFromPSPath(path, cmdContext);
+
+                foreach (PathInfo pathInfo in pathInfos)
```
If you resolve to multiple files, shouldn't that be an error instead of just ignoring it?

SteveL-MSFT replied:
Since wildcards are explicitly being checked and not allowed in the module manifest, is a Debug.Assert sufficient?

SteveL-MSFT replied:
looking at other code in this file, decided to throw

SteveL-MSFT replied:
Possible downside here is I'm adding blocks that can't be hit

SteveL-MSFT committed:
throw if resolving file path returns more than one result
added negative test cases that do file path checks

SteveL-MSFT reviewed (commented):
@lzybkr are you ok with the updated changes?

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,59 @@
+            }
+            catch {
+                $_.FullQulaifiedErrorId | Should Match "$errorId"
+            }
```
Why `Should BeOfType System.Management.Automation.PSModuleInfo` ? If we expect that ` Test-ModuleManifest` raise an exception in the tests the `Should BeOfType` will be never run.

I suggest using the excellent function [ShouldBeErrorId](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Common/Test.Helpers.psm1#L48)
(I have already prepared new PR to move existing tests to this pattern)
The new look of this test:
```powershell
            { Test-ModuleManifest -Path $testModulePath -ErrorAction Stop } | ShouldBeErrorId $errorId
```
Only insert:
```powershell
Import-Module $PSScriptRoot\..\..\Common\Test.Helpers.psm1
```

SteveL-MSFT replied:
Cool!  will make this change

iSazonov replied:
Closed.

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,59 @@
+Describe "Test-ModuleManifest tests" -tags "CI" {
+
+    AfterEach {
+        if (Test-Path testdrive:/module) {
```
It looks as unneeded check: `Remove-Item` implicitly do the check.

SteveL-MSFT replied:
changed

iSazonov replied:
Closed.

SteveL-MSFT commented:
@iSazonov thanks for the feedback, I'll take care of these after I get back from vacation on 2/21

SteveL-MSFT commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,59 @@
+        New-Item -ItemType Directory -Path testdrive:/module/foo
+        New-Item -ItemType Directory -Path testdrive:/module/bar
+        "" > testdrive:/module/foo/bar.psm1
+        "" > testdrive:/module/bar/foo.psm1
```
Agree.  Fixed.

iSazonov replied:
Closed.

SteveL-MSFT commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,59 @@
+
+        Test-Path $testModulePath | Should Be $true
+
+        Test-ModuleManifest -Path $testModulePath -ErrorAction Stop | Should BeOfType System.Management.Automation.PSModuleInfo
```
I used `ErrorAction Stop` here since Test-ModuleManifest writes to the error stream on manifest errors and I thought this was an easy way to ensure no errors being returned.  Otherwise I think I'd have to redirect the error stream and check it and I don't think that's providing additional value.  I can add a comment to clarify my intent.

iSazonov replied:
Clear.
Closed.

SteveL-MSFT commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,59 @@
+
+        New-Item -ItemType Directory -Path testdrive:/module
+        New-Item -ItemType Directory -Path testdrive:/module/foo
+        "" > testdrive:/module/foo/bar.psm1
```
Fixed

iSazonov replied:
Closed.

SteveL-MSFT commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,59 @@
+            "RootModule"="Modules_InvalidManifest";
+            "ScriptsToProcess"="Modules_InvalidManifest"
+        }
+        foreach ($parameter in $parametersAndExecptions.Keys) {
```
good catch

iSazonov replied:
Closed.

iSazonov commented:
@SteveL-MSFT I don't see new commit related with your last comments.

SteveL-MSFT committed:
changes to tests based on review feedback to make it simpler and more readable
fixed typo in test that exposed issue in cmdlet for validating nestedmodules

SteveL-MSFT commented:
@iSazonov took longer than I expected, after I fixed the typo, I found a issue with the cmdlet

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -37,22 +38,17 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
             "TypesToProcess"="Modules_InvalidManifest";
             "FormatsToProcess"="Modules_InvalidManifest";
-            "RootModule"="Modules_InvalidManifest";
+            #"RootModule"="Modules_InvalidManifest";
```
Maybe remove it?

SteveL-MSFT replied:
Should be uncommented.  Looks like the cmdlets doesn't do any validation of RootModule currently, will need to add that code.

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -37,22 +38,17 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
         }
-        foreach ($parameter in $parametersAndExecptions.Keys) {
+        foreach ($parameter in $parametersAndErrors.Keys) {
+            Write-Warning "Testing $parameter"
```
Remove?

SteveL-MSFT replied:
Not sure what the best practice is here.  When doing the test in a loop and something fails, the current error from Pester isn't sufficient to tell you where it failed.  @JamesWTruher suggestions?

iSazonov replied:
I suppose we should use `TestCases`

SteveL-MSFT replied:
Cool!

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -37,22 +38,17 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
+            [string]$errorId = $parametersAndErrors[$parameter] + ",Microsoft.PowerShell.Commands.TestModuleManifestCommand"
+
+            { Test-ModuleManifest -Path $testModulePath -ErrorAction Stop } | ShouldBeErrorId $errorId
             Remove-Item -Recurse -Force $testModulePath
```
Remove? It seems `AfterEach` does cleanups.

SteveL-MSFT replied:
This is needed as I'm generating a new invalid manifest each iteration of the loop

SteveL-MSFT committed:
fixed issue with validation of rootmodule
used TestCases feature of Pester for tests
added tests for valid and invalid rootmodules

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -24,32 +24,74 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
-            "ScriptsToProcess"="Modules_InvalidManifest"
+        $args = @{$parameter = "doesnotexist.psm1"}
+        New-ModuleManifest -Path $testModulePath @args
+        Test-Path $testModulePath | Should Be $true
```
Below `ShouldBeErrorId` still catches this error. Maybe remove the line?

SteveL-MSFT replied:
Yeah, this isn't needed, originally I thought to check that the manifest was actually created, but it would fail later.  Will remove.

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -24,32 +24,74 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
+        New-Item -ItemType Directory -Path testdrive:/module
+        $testModulePath = "testdrive:/module/test.psd1"
+
+        if ($rootModuleValue -ne $null -and $rootModuleValue -ne "")
```
It seems it is better to us to split the tests into two and remove the conditions.

SteveL-MSFT replied:
Was perhaps over optimizing on code reuse.  I'll split it to make it more readable.

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -24,32 +24,74 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
-            { Test-ModuleManifest -Path $testModulePath -ErrorAction Stop } | ShouldBeErrorId $errorId
-            Remove-Item -Recurse -Force $testModulePath
+        New-ModuleManifest -Path $testModulePath -RootModule $rootModuleValue
+        Test-Path $testModulePath | Should Be $true
```
Maybe remove too?

SteveL-MSFT replied:
Agree

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -24,32 +24,74 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
+        New-Item -ItemType Directory -Path testdrive:/module
+        $testModulePath = "testdrive:/module/test.psd1"
+
+        if ($rootModuleValue -ne "doesnotexist.psm1")
```
It seems it is better to us to split the tests into two and remove the conditions.

SteveL-MSFT replied:
I'm ok with splitting

iSazonov commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -24,32 +24,74 @@ Describe "Test-ModuleManifest tests" -tags "CI" {
+            New-Item -ItemType File -Path testdrive:/module/$rootModuleValue
+        }
+        New-ModuleManifest -Path $testModulePath -RootModule $rootModuleValue
+        Test-Path $testModulePath | Should Be $true
```
Maybe remove too?

SteveL-MSFT replied:
Agree

SteveL-MSFT committed:
split tests to make them more readable, removed unnecessary checks for manifest creation

iSazonov commented:
I see multiple `New-Item` but I see no reason optimize them (BeforeAll).

So the tests LGTM.

lzybkr commented on src/System.Management.Automation/engine/Modules/TestModuleManifestCommand.cs:
```diff
@@ -131,6 +131,22 @@ protected override void ProcessRecord()
                             }
                         }
 
+                        //RootModule can be null, empty string or point to a valid .psm1, or .dll.  Anything else is invalid.
```
`cdxml` and `xaml` modules are allowed in `RootModule` just as they are allowed in `NestedModules`.

lzybkr replied:
Also, when comparing extensions, is ignoring the case correct for non-Windows?

iSazonov replied:
This file contains `OrdinalIgnoreCase` 4 times. But I see lots of other files to ignore the case when dealing with paths.

lzybkr replied:
Yes, case insensitive comparisons on executable extensions could be a bigger issue than just this file.

SteveL-MSFT replied:
I think the extension comparison issue should probably be a separate issue so we can apply same fix everywhere

SteveL-MSFT replied:
Created issue https://github.com/PowerShell/PowerShell/issues/3218

iSazonov replied:
I agree. We don't even have the desired method. Perhaps even better, make a request in .Net.

SteveL-MSFT replied:
In testing a fix for cdxml/XAML modules, I see that you get this error message for InvalidModuleManifest which just seems wrong:

https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/src/System.Management.Automation/resources/Modules.resx#L157

    <value>The module manifest '{0}' could not be processed because it is not a valid Windows PowerShell restricted language file. Remove the elements that are not permitted by the restricted language: {1}</value>

I'd like to change it to something like:

The module manifest '{0}' could not be processed because it is not a valid PowerShell module manifest file.  Remove the elements that are not permitted: {1}

CC @HemantMahawar

SteveL-MSFT replied:
@lzybkr addressed

SteveL-MSFT committed:
added check for cdxml and xaml rootmodules which are allowed
added tests
updated InvalidModuleManifest error message to be about module manifests

SteveL-MSFT committed:
added check for cdxml and xaml rootmodules which are allowed
added tests
updated InvalidModuleManifest error message to be about module manifests
changed -TestCases usage to parens than backtick for readability

lzybkr commented on test/powershell/engine/Module/TestModuleManifest.ps1:
```diff
@@ -0,0 +1,121 @@
+        Test-ModuleManifest -Path $testModulePath -ErrorAction Stop | Should BeOfType System.Management.Automation.PSModuleInfo
+    }
+
+    It "module manifest containing missing files returns error" -TestCases `
```
Typical use of `-TestCases` is to introduce a variable - that is probably easier to read.

If there is a good reason to inline the argument - you should prefer parens instead of a backtick.
Extra indentation on the test cases might also make it more readable, so maybe format like:

```PowerShell
It "should work" -TestCases (
        @{rootModuleValue = "bar.psm1"},
        @{rootModuleValue = "bar.dll"}
) {
    param($rootModuleValue)
}
```

SteveL-MSFT replied:
will change

SteveL-MSFT committed:
Merge branch 'modulemanifest' of github.com:SteveL-MSFT/PowerShell into modulemanifest

lzybkr commented:
Closing and re-opening to retry blocked CI.
